src/wallet/bdb.cpp

```diff
@@ -668,9 +669,15 @@ bool BerkeleyCursor::Next(CDataStream& ssKey, CDataStream& ssValue, bool& comple
     complete = false;
     if (m_cursor == nullptr) return false;
     // Read at cursor
-    SafeDbt datKey;
+    SafeDbt datKey(m_key_prefix.data(), m_key_prefix.size());
```

Comment:
AFAICT it just overwrites the `datKey.data` with a pointer to newly allocated memory. There's nothing that indicates that this doesn't work (i.e. no tests fail and wallets load succesfully).

Reply:
> AFAICT it just overwrites the `datKey.data` with a pointer to newly allocated memory. There's nothing that indicates that this doesn't work (i.e. no tests fail and wallets load succesfully).

I don't see how it could allocate memory if the [DB_DBT_MALLOC flag is not set](https://github.com/bitcoin/bitcoin/blob/15bc973d0db48a8d5f44abfec4664dcf58de21e8/src/wallet/bdb.cpp#L223-L231)

I didn't get far enough in the PR to see if this has test coverage, but I guess you are pretty confident tests would fail if this was not working. It just seems like a mystery how it could work.

Reply:
Considering that many prefixes are shorter than entire keys, if there was a problem here, I would expect deserialization of keys to be throwing exceptions, and that would cause any test that calls `loadwallet` to fail, which would result in a ton of test failures. Given that that doesn't happen, I don't think there's any issue here.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/24914#discussion_r989523042

> I don't see how it could allocate memory if the [DB_DBT_MALLOC flag is not set](https://github.com/bitcoin/bitcoin/blob/15bc973d0db48a8d5f44abfec4664dcf58de21e8/src/wallet/bdb.cpp#L223-L231)

I went down a rabbithole trying to answer a question in another PR (https://github.com/bitcoin/bitcoin/pull/27224/files#r1150232826), and finally figured out how this could work...

It turns the DB_DBT_MALLOC flag is not a switch that controls whether BDB mallocs a new pointer for you or if it uses the pointer that you provide. BDB will malloc a new pointer whether or not you not pass DB_DBT_MALLOC. But you pass it, you are responsible for freeing the pointer, and if you don't pass it, BDB will manage the memory as part of the cursor object so you don't have to free it yourself. If you don't want BDB to malloc memory to return the key, you can pass it a different DB_DBT_USERMEM flag instead.

This does raises the question of whether it is ok to call memory_cleanse on memory that is owned by the cursor and should probably be read-only. But maybe it doesn't need to be read-only, or doesn't need to be read-only as long as you are done with the key and about to advance the cursor.